HW2/P3/mandelbrot.pyx

```diff
@@ -42,7 +42,88 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
                     z = z * z + c
                 out_counts[i, j] = iter
 
+# use instruction level
+cdef void counts_to_output(AVX.float8 counts,
+                           np.uint32_t [:, :] out_counts,
+                           int i,
+                           int j) nogil:
+    cdef:
+        float tmp_counts[8]
+        int k
+
+    AVX.to_mem(counts, &(tmp_counts[0]))
+
+    for k in range(8):
+        out_counts[i, j*8+k]=<unsigned int>tmp_counts[k]
+
+cpdef mandelbrot_new(np.complex64_t [:, :] in_coords,
+                     np.uint32_t [:, :] out_counts,
+                     int max_iterations=511):
+    cdef:
+       int i, j, iter, k
+       float [:, :] real, imag
+       AVX.float8 cr, ci, zr, zr_n, zi, magz2, signs, counts
+
+    real = np.real(in_coords)
+    imag = np.imag(in_coords)
+
+
+       # You may find the numpy.real() and numpy.imag() fuctions helpful.
+    assert in_coords.shape[1] % 8 == 0, "Input array must have 8N columns"
+    assert in_coords.shape[0] == out_counts.shape[0], "Input and output arrays must be the same size"
+    assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
+
+    with nogil:
+        for i in prange(in_coords.shape[0], num_threads=4, schedule='static', chunksize=1):
+            for j in range(in_coords.shape[1]/8):
+                cr = AVX.make_float8(real[i,j*8+7],
+                                     real[i,j*8+6],
+                                     real[i,j*8+5],
+                                     real[i,j*8+4],
+                                     real[i,j*8+3],
+                                     real[i,j*8+2],
+                                     real[i,j*8+1],
+                                     real[i,j*8])
+
+                ci = AVX.make_float8(imag[i,j*8+7],
+                                     imag[i,j*8+6],
+                                     imag[i,j*8+5],
+                                     imag[i,j*8+4],
+                                     imag[i,j*8+3],
+                                     imag[i,j*8+2],
+                                     imag[i,j*8+1],
+                                     imag[i,j*8])
+
+                zr = AVX.float_to_float8(0.0)
+                zi = AVX.float_to_float8(0.0)
+                counts = AVX.float_to_float8(0.0)
+
+                for iter in range(max_iterations):
+
+                    # @x: calculate the magnitude square
+                    magz2 = AVX.add(AVX.mul(zr,zr), AVX.mul(zi,zi))
+
+                    # @x: check if magnitude square of z less than 4, return 1 if true, return 0 if false
+                    signs = AVX.bitwise_and(AVX.less_than(magz2, AVX.float_to_float8(4.0)),AVX.float_to_float8(1.0))
```

Comment:
@isadoranun I think checking for magnitude >= 4 or magnitude > 4 makes no difference in practice. The chances converge against zero that the case == 4 will ever be reached (thx. to floating point arithmetic), and even if so the 'error' of one false iteration count is neglectable as our coloring is done absolutely arbitrarily. I.e. the assumption that the series converges only if the result after 511 iterations is still inside an arbitrary bound is not really strict. There might be cases where we find out that the bound may be passed in the 512 iteration.

:+1: for mask